src/weighted_list.h

```diff
@@ -124,41 +137,87 @@ template <typename W, typename T> struct weighted_list {
     /**
      * This will return the sum of all the object's weights in the list.
      */
-    double get_weight() const {
+    W get_weight() const {
         return total_weight;
     }
 
+    typedef std::vector<weighted_object<W, T> > weighted_object_list;
+    typedef typename weighted_object_list::iterator iterator;
+    typedef typename weighted_object_list::const_iterator const_iterator;
+    iterator begin() { return objects.begin(); }
+    iterator end() { return objects.end(); }
+    iterator erase (const_iterator first, const_iterator last) {
+        invalidate_precalc();
+        return objects.erase(first, last);
+    };
+    size_t size() const noexcept { return objects.size(); }
+    bool empty() const noexcept { return objects.empty(); }
+
+    void precalc();
+
 protected:
     W total_weight;
-    std::vector<weighted_object<W, T> > objects;
+    weighted_object_list objects;
 
-    virtual size_t pick_ent() const =0;
+    virtual size_t pick_ent(long long) const =0;
+    void invalidate_precalc() {}
 };
 
 template <typename T> struct weighted_int_list : public weighted_list<int, T> {
 
+    // populate the precalc_array for O(1) lookups
+    void precalc(){
+        precalc_array.clear();
+        precalc_array.reserve(this->total_weight); // to avoid additional reallocations
+        size_t i;
+        // weights [3,1,5] will produce vector of indices [0,0,0,1,2,2,2,2,2]
+        for(i=0; i < this->objects.size(); i++) {
+            precalc_array.resize(precalc_array.size() + this->objects[i].weight, i);
+        }
+    }
+
 protected:
 
-    size_t pick_ent() const override {
-        int picked = rng(1, this->total_weight);
-        int accumulated_weight = 0;
+    size_t pick_ent(long long rand_ll) const override {
         size_t i;
-        for(i=0; i < this->objects.size(); i++) {
-            accumulated_weight += this->objects[i].weight;
-            if(accumulated_weight >= picked) {
-                break;
+        int picked;
+        picked = (rand_ll%(this->total_weight))+1;
```

Comment:
Combine this and the line above into one `int picked = ...;` - no point in having those as separate statements.

Btw. if you make `precalc_array` mutable, you can change it in this (const) function. Which means you can re-calculated it here, when you need it for the first time. Consequently, you don't need the `precalc` function. It would also allow you to bypass it when the list contains only one entry.

Just a sketch:
```C++
{
    if( objects.size() == 1 ) {
        return 0;
    }
    if( precalcs.empty() ) {
        // fill precalcs with proper values
    }
    return precalcs[...]
}
```